src/commands/tail.rs

```diff
@@ -1,132 +1,76 @@
-use std::net::{SocketAddr, TcpListener};
-
-use anyhow::Result;
-
+use crate::http;
 use crate::settings::global_user::GlobalUser;
 use crate::settings::toml::Target;
-use crate::tail::Tail;
-
-const DEFAULT_TUNNEL_PORT: u16 = 8080;
-const DEFAULT_METRICS_PORT: u16 = 8081;
 
-pub fn start(
-    target: &Target,
-    user: &GlobalUser,
-    format: String,
-    tunnel_port: Option<u16>,
-    metrics_port: Option<u16>,
-    verbose: bool,
-) -> Result<()> {
-    let tunnel_port = find_open_port(tunnel_port, DEFAULT_TUNNEL_PORT)?;
-    let metrics_port = find_open_port(metrics_port, DEFAULT_METRICS_PORT)?;
-
-    Tail::run(
-        target.clone(),
-        user.clone(),
-        format,
-        tunnel_port,
-        metrics_port,
-        verbose,
-    )
+/// `wrangler tail` allows Workers users to collect logs from their deployed Workers.
+/// When a user runs `wrangler tail`, several things happen:
+///     1. wrangler asks the Cloudflare API to send log events to a Durable Object, which will act as a log forwarder.
+///        The API returns the address of the log forwarder.
+///     3. wrangler connects to the log forwarder with a Websocket.
+///     5. Upon receipt, wrangler prints log events to STDOUT.
+use anyhow::Result;
+use cloudflare::endpoints::workers::{CreateTail, CreateTailParams};
+use cloudflare::framework::apiclient::ApiClient;
+use futures::stream::StreamExt;
+
+#[derive(serde::Deserialize, Debug)]
+struct TailResult {
+    id: String,
+    expires_at: String,
 }
 
-/// Find open port takes two arguments: an Optional requested port, and a default port.
-fn find_open_port(requested: Option<u16>, default: u16) -> Result<u16> {
-    if let Some(port) = requested {
-        let addr = format!("127.0.0.1:{}", port);
-        match TcpListener::bind(addr) {
-            Ok(socket) => Ok(socket.local_addr()?.port()),
-            Err(_) => anyhow::bail!("the specified port {} is unavailable", port),
+// Main loop, listen for websocket messages or interrupt
+async fn listen_tail(tail_tag: String) -> Result<(), ()> {
+    // ws listener setup
+    let listen_tail_endpoint = format!("wss://tail.developers.workers.dev/{}/ws", tail_tag);
```

Comment:
Just occurred to me that we may want to return the full URL in the API response. This lets us change it more easily in the future, though I can't think of a reason we would actually need to. Could aid with testing at least.